packages/sitecore-jss-react/src/components/PlaceholderMetadata.test.tsx

```diff
@@ -0,0 +1,109 @@
+
+    const actualOutput = normalizeWhitespace(wrapper.html());
+    const normalizedExpectedOutput = normalizeWhitespace(expectedOutput);
+
+    expect(actualOutput).to.equal(normalizedExpectedOutput);
+  });
+
+  it('renders deeply nested components correctly', () => {
+    const component = {
+      uid: 'root123',
+      componentName: 'Layout',
+      placeholders: {
+        header: [
+          {
+            uid: 'nested123',
+            componentName: 'Header',
+            placeholders: {
+              logo: [
+                {
+                  uid: 'deep123',
+                  componentName: 'Logo',
+                },
+              ],
+            },
+          },
+        ],
+      },
+    };
+
+    const wrapper = mount(<PlaceholderMetadata rendering={component} />);
+
+    const expectedOutput =
+      '<code type="text/sitecore" chrometype="placeholder" class="scpm" kind="open" id="header_root123"></code><code type="text/sitecore" chrometype="rendering" class="scpm" kind="open" id="nested123"></code><header><code type="text/sitecore" chrometype="placeholder" class="scpm" kind="open" id="logo_nested123"></code><code type="text/sitecore" chrometype="rendering" class="scpm" kind="open" id="deep123"></code><logo></logo><code type="text/sitecore" chrometype="rendering" class="scpm" kind="close" id="deep123"></code><code type="text/sitecore" chrometype="placeholder" class="scpm" kind="close" id="logo_nested123"></code></header><code type="text/sitecore" chrometype="rendering" class="scpm" kind="close" id="nested123"></code><code type="text/sitecore" chrometype="placeholder" class="scpm" kind="close" id="header_root123"></code>';
+
+    const actualOutput = normalizeWhitespace(wrapper.html());
+    const normalizedExpectedOutput = normalizeWhitespace(expectedOutput);
+
+    expect(actualOutput).to.equal(normalizedExpectedOutput);
+  });
+
+  it('should render component if nested placeholder is empty', () => {
+    const component = {
+      uid: '123',
+      componentName: 'Layout',
+      placeholders: {
+        main: [],
+      },
+    };
+
+    const wrapper = mount(<PlaceholderMetadata rendering={component} />);
+
+    const expectedOutput =
+      '<code type="text/sitecore" chrometype="placeholder" class="scpm" kind="open" id="main_123"></code><code type="text/sitecore" chrometype="placeholder" class="scpm" kind="close" id="main_123"></code>';
+
+    const actualOutput = normalizeWhitespace(wrapper.html());
+    const normalizedExpectedOutput = normalizeWhitespace(expectedOutput);
+
+    expect(actualOutput).to.equal(normalizedExpectedOutput);
+  });
+});
```

Comment:
Please also add a test case for placeholder that renders a component with missing implementation (not registered in componentBuilder).
This component should still be wrapped in placeholder and rendering code blocks.